engine/math/matrix.lisp

```diff
@@ -0,0 +1,545 @@
+		   (y value))))
+   (reduce #'+
+	   (loop
+	      for i
+	      from 0
+	      to 2 collect
+		(* (get-at matrix 2 i)
+		   (z value))))))
+
+(defmethod mul ((matrix float4x4) (value float4))
+  (make-float4
+   (reduce #'+
+	   (loop
+	      for i
+	      from 0
+	      to 3 collect
+		(* (get-at matrix 0 i)
+		   (x value))))
+   (reduce #'+
+	   (loop
+	      for i
+	      from 0
+	      to 3 collect
+		(* (get-at matrix 1 i)
+		   (y value))))
+   (reduce #'+
+	   (loop
+	      for i
+	      from 0
+	      to 3 collect
+		(* (get-at matrix 2 i)
+		   (z value))))
+   (reduce #'+
+	   (loop
+	      for i
+	      from 0
+	      to 3 collect
+		(* (get-at matrix 3 i)
+		   (w value))))))
+
+
+; small helper functions for determinants, maybe should be macros
+(defun det2x2 (a11 a12 a21 a22)
+  (- (* a11 a22)
+     (* a12 a21)))
+
+(defun det3x3 (a11 a12 a13 a21 a22 a23 a31 a32 a33)
+  (- (+ (* a11
+	   (det2x2 a22 a23
+		   a32 a33))
+	(* a13
+	   (det2x2 a21 a22
+		   a31 a32)))
+     (* a12
+	(det2x2 a21 a23
+		a31 a33))))
+
+(defun det4x4 (a11 a12 a13 a14 a21 a22 a23 a24 a31 a32 a33 a34 a41 a42 a43 a44)
+  (- (+ (* a11
+	   (det3x3 a22 a23 a24
+		   a32 a33 a34
+		   a42 a43 a44))
+	(* a13
+	   (det3x3 a21 a22 a24
+		   a31 a32 a34
+		   a41 a42 a44)))
+     (+ (* a12
+	   (det3x3 a21 a23 a24
+		   a31 a33 a34
+		   a41 a43 a44))
+	(* a14
+	   (det3x3 a21 a22 a23
+		   a31 a32 a33
+		   a41 a42 a43)))))
+
+; should be macros too I think
+(defmethod det ((matrix float2x2))
+  (det2x2 (get-at matrix 0 0)
+	  (get-at matrix 0 1)
+	  (get-at matrix 1 0)
+	  (get-at matrix 1 1)))
+
+(defmethod det ((matrix float3x3))
+  (det3x3 (get-at matrix 0 0)
+	  (get-at matrix 0 1)
+	  (get-at matrix 0 2)
+	  (get-at matrix 1 0)
+	  (get-at matrix 1 1)
+	  (get-at matrix 1 2)
+	  (get-at matrix 2 0)
+	  (get-at matrix 2 1)
+	  (get-at matrix 2 2)))
+
+(defmethod det ((matrix float4x4))
+  (det4x4 (get-at matrix 0 0)
+	  (get-at matrix 0 1)
+	  (get-at matrix 0 2)
+	  (get-at matrix 0 3)
+	  (get-at matrix 1 0)
+	  (get-at matrix 1 1)
+	  (get-at matrix 1 2)
+	  (get-at matrix 1 3)
+	  (get-at matrix 2 0)
+	  (get-at matrix 2 1)
+	  (get-at matrix 2 2)
+	  (get-at matrix 2 3)
+	  (get-at matrix 3 0)
+	  (get-at matrix 3 1)
+	  (get-at matrix 3 2)
+	  (get-at matrix 3 3)))
+
+
+(defmethod absm ((matrix float2x2))
+  (make-float2x2
+   (map 'vector
+	#'abs
+	(in-list matrix))))
+
+(defmethod absm ((matrix float3x3))
+  (make-float3x3
+   (map 'vector
+	#'abs
+	(in-list matrix))))
+
+(defmethod absm ((matrix float4x4))
+  (make-float4x4
+   (map 'vector
+	#'abs
+	(in-list matrix))))
+
+
+(defmethod negm ((matrix float2x2))
+  (make-float2x2
+   (map 'vector
+	#'-
+	(in-list matrix))))
+
+(defmethod negm ((matrix float3x3))
+  (make-float3x3
+   (map 'vector
+	#'-
+	(in-list matrix))))
+
+(defmethod negm ((matrix float4x4))
+  (make-float4x4
+   (map 'vector
+	#'-
+	(in-list matrix))))
+
+
+; helper function for transpose
+(defun flatten (l)
```

Comment:
I think we should use [alexandria](https://common-lisp.net/project/alexandria/draft/alexandria.html#Sequences) flatten function instead.